include/oneapi/dpl/pstl/execution_impl.h

```diff
@@ -29,100 +30,132 @@ namespace dpl
+//----------------------------------------------------------
+
+template <class _Tag>
+struct __is_backend_tag_parallel_forward : ::std::false_type
 {
-    typedef ::std::true_type __allow_parallel;
-    typedef ::std::false_type __allow_unsequenced;
-    typedef ::std::false_type __allow_vector;
 };
 
 template <>
-struct __policy_traits<oneapi::dpl::execution::parallel_unsequenced_policy>
+struct __is_backend_tag_parallel_forward<__parallel_forward_tag> : ::std::true_type
 {
-    typedef ::std::true_type __allow_parallel;
-    typedef ::std::true_type __allow_unsequenced;
-    typedef ::std::true_type __allow_vector;
 };
 
-template <typename _ExecutionPolicy>
-using __allow_vector = typename __internal::__policy_traits<::std::decay_t<_ExecutionPolicy>>::__allow_vector;
-
-template <typename _ExecutionPolicy>
-using __allow_unsequenced = typename __internal::__policy_traits<::std::decay_t<_ExecutionPolicy>>::__allow_unsequenced;
+template <class _Tag>
+inline constexpr bool __is_backend_tag_parallel_forward_v =
+    __is_backend_tag_parallel_forward<::std::decay_t<_Tag>>::value;
 
-template <typename _ExecutionPolicy>
-using __allow_parallel = typename __internal::__policy_traits<::std::decay_t<_ExecutionPolicy>>::__allow_parallel;
+//----------------------------------------------------------
+// __is_backend_tag_parallel, __is_backend_tag_parallel_v
+//----------------------------------------------------------
 
-template <typename _ExecutionPolicy, typename... _IteratorTypes>
-auto
-__is_vectorization_preferred(_ExecutionPolicy& __exec)
-    -> decltype(__internal::__lazy_and(__exec.__allow_vector(),
-                                       __internal::__is_random_access_iterator_t<_IteratorTypes...>()))
+template <class _Tag>
+struct __is_backend_tag_parallel : ::std::false_type
 {
-    return __internal::__lazy_and(__exec.__allow_vector(),
-                                  __internal::__is_random_access_iterator_t<_IteratorTypes...>());
-}
+};
 
-template <typename _ExecutionPolicy, typename... _IteratorTypes>
-auto
-__is_parallelization_preferred(_ExecutionPolicy& __exec)
-    -> decltype(__internal::__lazy_and(__exec.__allow_parallel(),
-                                       __internal::__is_random_access_iterator_t<_IteratorTypes...>()))
+template <class _IsVector>
+struct __is_backend_tag_parallel<__parallel_tag<_IsVector>> : ::std::true_type
 {
-    return __internal::__lazy_and(__exec.__allow_parallel(),
-                                  __internal::__is_random_access_iterator_t<_IteratorTypes...>());
-}
+};
+
+template <class _Tag>
+inline constexpr bool __is_backend_tag_parallel_v = __is_backend_tag_parallel<::std::decay_t<_Tag>>::value;
+
+//----------------------------------------------------------
+// __is_host_backend_tag_v
+//----------------------------------------------------------
+
+template <class _Tag>
+inline constexpr bool __is_host_backend_tag_v =
```

Comment:
Overall comment. I don't understand logic how we choose what static_assert condition to use. Also I don't understand, why do we need `__is_host_backend_tag_v`?

Reply:
The idea was very simple.
For example, let's find what we have now for `__pattern_walk2_transform_if(`
- we have one host impl and one hetero impl:
```C++
template <class _Tag, typename _ExecutionPolicy, typename _ForwardIterator1, typename _ForwardIterator2,
          typename _Function>
_ForwardIterator2
__pattern_walk2_transform_if(_Tag __tag, _ExecutionPolicy&& __exec, _ForwardIterator1 __first1,
                             _ForwardIterator1 __last1, _ForwardIterator2 __first2, _Function __func) noexcept
{
    static_assert(__is_host_backend_tag_v<_Tag>);

    return __pattern_walk2(__tag, ::std::forward<_ExecutionPolicy>(__exec), __first1, __last1, __first2, __func);
}
```

```C++
template <typename _BackendTag, typename _ExecutionPolicy, typename _ForwardIterator1, typename _ForwardIterator2,
          typename _Function>
_ForwardIterator2
__pattern_walk2_transform_if(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _ForwardIterator1 __first1,
                             _ForwardIterator1 __last1, _ForwardIterator2 __first2, _Function __func)
{
    // Require `read_write` access mode for output sequence to force a copy in for host iterators to capture incoming
    // values of the output sequence for elements where the predicate is false.
    return __pattern_walk2<_BackendTag, /*_IsSync=*/::std::true_type, __par_backend_hetero::access_mode::read,
                           __par_backend_hetero::access_mode::read_write>(
        __tag,
        __par_backend_hetero::make_wrapped_policy<__walk2_transform_if_wrapper>(
            ::std::forward<_ExecutionPolicy>(__exec)),
        __first1, __last1, __first2, __func);
}
```

So What is helpful to check in the host impl? I check that the tag is any host tag and if I have hetero tag here it's compile error.